### PR TITLE
LLM-323/embeddings content

### PR DESCRIPTION
**Pull request by NickSeagull**
Save content as metadata.

I also added a Nix dev environment so one can get an instant dev environment by `cd`-ing into the root project.

### CODE REVIEW

File: modules/semanticsearch/src/main/kotlin/com/theagilemonkeys/ellmental/semanticsearch/SemanticSearch.kt [hunk 1]
```diff
@@ -15,38 +15,38 @@ context(EmbeddingsModel<Any>, VectorStore)
...
-        check(text.isNotBlank()) { "Text cannot be blank" }
-        val embedding = embed(text)
-        val semanticEntry = SemanticEntry(embedding = embedding)
-        upsert(semanticEntry)
-    }
+    suspend fun learn(input: SearchInput) =
+        input.texts.forEach { text ->
+            check(text.isNotBlank()) { "Text cannot be blank" }
+            val embedding = embed(text)
+            val semanticEntry = SemanticEntry(content = text, embedding = embedding)
```
Comment (NickSeagull) [inline modules/semanticsearch/src/main/kotlin/com/theagilemonkeys/ellmental/semanticsearch/SemanticSearch.kt:28]: This line was the change

File: modules/vectorstore/src/main/kotlin/com/theagilemonkeys/ellmental/vectorstore/pinecone/PineconeVectorStore.kt [hunk 2]
```diff
@@ -47,14 +48,18 @@ class PineconeVectorStore(
...
-            Vectors(
-                id = semanticEntry.id.value,
-                values = semanticEntry.embedding.value,
-                sparseValues = null,
-                metadata = semanticEntry.metadata
-            ),
-        )
+        val metadata1 = semanticEntry.metadata ?: emptyMap()
+        val metadata2 = mapOf("content" to semanticEntry.content)
+        val newMetadata = metadata1 + metadata2
```
Comment (NickSeagull) [inline modules/vectorstore/src/main/kotlin/com/theagilemonkeys/ellmental/vectorstore/pinecone/PineconeVectorStore.kt:53]: Here I'm adding the content to the metadata

File: modules/vectorstore/src/main/kotlin/com/theagilemonkeys/ellmental/vectorstore/pinecone/PineconeVectorStore.kt [hunk 3]
```diff
@@ -66,23 +71,25 @@ class PineconeVectorStore(
...
-                embedding = Embedding(it.values ?: emptyList()),
-                metadata = it.metadata ?: emptyMap(),
-            )
-        }
+        val response = Json.decodeFromString(Schema.QueryResponse::class.toSerializer(), responseString)
+        val entries =
+            response.matches.map {
+                SemanticEntry(
+                    id = Id(it.id),
+                    content = it.metadata?.get("content") ?: "",
```
Comment (NickSeagull) [inline modules/vectorstore/src/main/kotlin/com/theagilemonkeys/ellmental/vectorstore/pinecone/PineconeVectorStore.kt:88]: Here I extract the metadata into the content

File: modules/vectorstore/src/main/kotlin/com/theagilemonkeys/ellmental/vectorstore/pinecone/PineconeVectorStore.kt (hunk 2 above)
Comment (javiertoledo) [reply modules/vectorstore/src/main/kotlin/com/theagilemonkeys/ellmental/vectorstore/pinecone/PineconeVectorStore.kt:53]: Are there any size limitations for the metadata in Pinecone? Would we say it's "safe" to store the chunks here?

File: shell.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+{ pkgs ? import <nixpkgs> {} }:
+pkgs.mkShell rec {
+  buildInputs = with pkgs; [
+    gradle # Not really needed, but just in case
+    kotlin-language-server
+    jdk17
+    nodejs # For building the website
+    jetbrains.idea-ultimate # IDE
+  ];
+}
```
Comment (javiertoledo) [inline shell.nix:30]: If we add this, it'd be nice to add a documentation section on "how to run this with nix"

**Review by javiertoledo — APPROVED** [review]
LGTM, just left a couple of notes

File: modules/vectorstore/src/main/kotlin/com/theagilemonkeys/ellmental/vectorstore/pinecone/PineconeVectorStore.kt (hunk 2 above)
Comment (NickSeagull) [reply modules/vectorstore/src/main/kotlin/com/theagilemonkeys/ellmental/vectorstore/pinecone/PineconeVectorStore.kt:53]: Each row has a limit of 40k of metadata in pinecone. I'd say it's safe, but who knows if someone will store a deeply nested JSON there... :trollface:

**Pull request merged**